src/Experimental/degeneracy.jl

```diff
@@ -0,0 +1,113 @@
+"""
+function PKC(g::AbstractSimpleGraph{T}, frac = 0.95) where T
+    has_self_loops(g) && throw(ArgumentError("graph must not have self-loops"))
+    deg = Atomic{T}.(degree(g))
+    buff = [Vector{T}() for _ in 1:nthreads()]
+    buff_end = zeros(Int64, nthreads())
+    visited = 0
+    level = 0
+    while visited < nv(g)*frac
+        visited += process_level(g, deg, level, buff, buff_end)
+        level += 1
+    end
+    if visited < nv(g)
+        g_small, vmap = subgraph(g, deg, level, nv(g)-visited)
+        U = eltype(g_small)
+        newdeg = Atomic{U}.(degree(g_small))
+        newbuff = [Vector{U}() for _ in 1:nthreads()]
+        while visited < nv(g)
+            visited += process_level(g_small, newdeg, level, newbuff, buff_end)
+            level += 1
+        end
+        @threads for i in 1:nv(g_small)
+            deg[vmap[i]][] = newdeg[i][]
+        end
+    end
+    core_num = Vector{T}(undef, nv(g))
+    @threads for v in 1:nv(g)
+        core_num[v] = deg[v][]
+    end
+    return core_num
+end
+
+function process_level(g::AbstractGraph{T}, deg::Vector{Atomic{T}}, level::Int64,
+                       buff::Vector{Vector{T}}, buff_end::Vector{Int64}) where T
+    buff_end .= 0
+    @threads for v in 1:nv(g)
+        if deg[v][] == level
+            tid = threadid()
+            push!(buff[tid], v)
+            buff_end[tid] += 1
+        end
+    end
+    @threads for tid in 1:nthreads()
+        while !isempty(buff[tid])
+            v = popfirst!(buff[tid])
+            for u in all_neighbors(g, v)
+                if deg[u][] > level
+                    a = atomic_sub!(deg[u], one(T))
+                    if a == level+1
+                        push!(buff[tid], u)
+                        buff_end[tid] += 1
+                    end
+                    if a <= level
+                        atomic_add!(deg[u], one(T))
+                    end
+                end
+            end
+        end
+    end
+    return sum(buff_end)
+end
+
+function subgraph(g::AbstractSimpleGraph{T}, deg::Vector{Atomic{T}}, level::Int64,
+                  nvg_small::Int64) where T
+    U = UInt8
+    while nvg_small > typemax(U)
+        U = widen(U)
+    end
+    g_small = SmallGraph{U}(nvg_small)
+    in_gsmall = falses(nv(g))
+    @threads for v in 1:nv(g)
+        if deg[v][] >= level
+            in_gsmall[v] = true
+        end
+    end
+    vmap = findall(in_gsmall)
+    newvid = Vector{Int64}(undef, nv(g))
+    @threads for i in 1:length(vmap)
+        newvid[vmap[i]] = i
+    end
+    @threads for s in vmap
+        for d in all_neighbors(g, s)
+            if in_gsmall[d]
+                add_edge!(g_small, newvid[s], newvid[d])
+            end
+        end
+    end
+    return g_small, vmap
+end
+
+#parallel graph
+struct SmallGraph{T <: Integer} <: AbstractGraph{T}
+    adjlist::Vector{Vector{T}}
+    nvg::T
+    SmallGraph{T}(nvertices) where T = new{T}([T[] for _ in 1:nvertices], nvertices)
+end
```

Comment:
What is the purpose of this `SmallGraph`? It looks very similar to a  `SimpleGraph` to me, but here we store the number of vertices instead of edges. (and the number of vertices seems to be fixed)

Reply:
It's used for creating a graph in parallel to avoid race conditions. SimpleGraphs won't work for building a graph in parallel, so I created this. I was just using it for testing, but I need a proper way to do this.

Reply:
why can't you just use multiple simplegraphs? `SmallGraph{T}` looks like it has the exact same structure as `SimpleGraph{T}` and that constructor looks pretty much identical.

Reply:
The difference is in the `add_edge!` function. In SimpleDiGraph when you add an edge u->v. It gets added to the `fadjlist` of u & `badjlist` of v, this can cause a race condition when another vertex is adding to v's `badjlist` in parallel. In this case when you call `add_edge!(g, u, v)`, v is only pushed to u's `adjlist`. It works well for this particular case. The same logic goes for SimpleGraphs, when an edge u->v is added in SimpleGraphs it gets pushed to both u & v's adjlist which can also cause a race condition.

Reply:
you're going to run into trouble here though, since you're not maintaining the adjacency lists in sorted order, which means that `outneighbors` is not sorted. This is bad for cache locality and branch prediction, and probably has other nasty side effects.

Reply:
Actually it works well for this case. Here we're creating a subgraph from a SimpleGraph, so when we push it to the adjacency list, they always go in sorted order. I'm not saying this will work in general, I just created this for a particular use case.

Reply:
Ah, ok. Let's add a comment here that nobody should ever use this outside of this function :)

Reply:
Scratch this, I'm using SimpleGraph now. I'm not using the `add_edge!` function, instead pushing to `fadjlist`. It's a hack but prevents creating a new structure.